File: timApp/modules/savearea/savearea.py
"""
TIM example plugin: a saveareandrome checker.
"""
import re
import sys
from typing import Union

import attr
from flask import jsonify, render_template_string
from marshmallow import Schema, fields, post_load, validates, ValidationError
from marshmallow.utils import missing
from webargs.flaskparser import use_args

from pluginserver_flask import GenericMarkupModel, GenericMarkupSchema, GenericHtmlSchema, GenericHtmlModel, \
    GenericAnswerSchema, GenericAnswerModel, Missing, \
    InfoSchema, create_app

sys.path.insert(0, '/py')  # /py on mountattu docker kontissa /opt/tim/timApp/modules/py -hakemistoon

@attr.s(auto_attribs=True)
class SaveareaStateModel:
    """Model for the information that is stored in TIM database for each answer."""
    demopisteet: float


class SaveareaStateSchema(Schema):
    demopisteet = fields.Number(required=True)

    @post_load
    def make_obj(self, data):
        return SaveareaStateModel(**data)

    class Meta:
        strict = True


@attr.s(auto_attribs=True)
class SaveareaMarkupModel(GenericMarkupModel):
    points_array: Union[str, Missing] = missing
    inputstem: Union[str, Missing] = missing
    needed_len: Union[int, Missing] = missing
    initword: Union[str, Missing] = missing
    cols: Union[int, Missing] = missing
    inputplaceholder: Union[str, Missing] = missing


class SaveareaMarkupSchema(GenericMarkupSchema):
    points_array = fields.List(fields.List(fields.Number()))
    inputstem2 = fields.Number()
    initword2 = fields.Number()
    cols = fields.Int()
    inputplaceholder2: Union[int, Missing] = missing

    @validates('points_array')
    def validate_points_array(self, value):
        if len(value) != 2 or not all(len(v) == 2 for v in value):
            raise ValidationError('Must be of size 2 x 2.')

    @post_load
    def make_obj(self, data):
        return SaveareaMarkupModel(**data)

    class Meta:
        strict = True


@attr.s(auto_attribs=True)
class SaveareaInputModel:
    """Model for the information that is sent from browser (plugin AngularJS component)."""
    demopisteet: float
    nosave: bool = missing


class SaveareaInputSchema(Schema):
    demopisteet = fields.Number(required=True)
    nosave = fields.Bool()

    @validates('demopisteet')
    def validate_demopisteet(self, number):
        if not number:
            raise ValidationError('Must be a number.')

    @post_load
    def make_obj(self, data):
        return SaveareaInputModel(**data)


class SaveareaAttrs(Schema):
    """Common fields for HTML and answer routes."""
    markup = fields.Nested(SaveareaMarkupSchema)
    state = fields.Nested(SaveareaStateSchema, allow_none=True, required=True)

    class Meta:
        strict = True


@attr.s(auto_attribs=True)
class SaveareaHtmlModel(GenericHtmlModel[SaveareaInputModel, SaveareaMarkupModel, SaveareaStateModel]):
    def get_component_html_name(self) -> str:
        return 'savearea-runner'

    def get_static_html(self) -> str:
        return render_static_savearea(self)

    def get_browser_json(self):
        r = super().get_browser_json()
        if self.state:
            r['demopisteet'] = self.state.demopisteet
        return r

    class Meta:
        strict = True


class SaveareaHtmlSchema(SaveareaAttrs, GenericHtmlSchema):
    info = fields.Nested(InfoSchema, allow_none=True, required=True)

    @post_load
    def make_obj(self, data):
        # noinspection PyArgumentList
        return SaveareaHtmlModel(**data)

    class Meta:
        strict = True


@attr.s(auto_attribs=True)
class SaveareaAnswerModel(GenericAnswerModel[SaveareaInputModel, SaveareaMarkupModel, SaveareaStateModel]):
    pass


class SaveareaAnswerSchema(SaveareaAttrs, GenericAnswerSchema):
    input = fields.Nested(SaveareaInputSchema, required=True)

    @post_load
    def make_obj(self, data):
        # noinspection PyArgumentList
        return SaveareaAnswerModel(**data)

    class Meta:
        strict = True


def render_static_savearea(m: SaveareaHtmlModel):
    return render_template_string(
        """
<div class="csRunDiv no-popup-menu">
    <h4>{{ header }}</h4>
    <p class="stem">{{ stem }}</p>
    <div><label>{{ inputstem or '' }} <span>
        <input type="text"
               class="form-control"
               placeholder="{{inputplaceholder}}"
               size="{{cols}}"></span></label>
    </div>
    <button class="timButton">
        {{ buttonText or button or "Save" }}
    </button>
    <a>{{ resetText }}</a>
    <p class="plgfooter">{{ footer }}</p>
</div>
        """,
        **attr.asdict(m.markup),
    )


app = create_app(__name__, SaveareaHtmlSchema())


@app.route('/answer/', methods=['put'])
@use_args(SaveareaAnswerSchema(strict=True), locations=("json",))
def answer(args: SaveareaAnswerModel):
    web = {}
    result = {'web': web}
    demopisteet = args.input.demopisteet

    return jsonify(result)


@app.route('/reqs/')
@app.route('/reqs')
def reqs():
    templates = ["""
``` {#ekasavearea plugin="savearea"}
header: Luo lomake
stem: Anna lomakkeen nimi ja kenttien lukumäärä.
-points_array: [[0, 0.1], [0.6, 1]]
inputstem: "Lomakkeen nimi:"
inputstem2: "Lomakkeen kenttien lukumäärä:"
needed_len: 1
initword: nimi
cols: 20
```""", """
``` {#tokasavearea plugin="savearea"}
header: Demopisteet
stem: Anna lomakkeen nimi ja kenttien lukumäärä.
-points_array: [[0, 0.1], [0.6, 1]]
inputstem2: 0
needed_len: 1
initword: nimi
cols: 20
```"""]
    return jsonify({
        "js": ["js/build/savearea.js"],
        "multihtml": True,
        "css": ["css/savearea.css"],
        'editor_tabs': [
            {
                'text': 'Plugins',
                'items': [
                    {
                        'text': 'Savearea:',
                        'items': [
                            {
                                'data': templates[0].strip(),
                                'text': 'Lomake ',
                                'expl': 'Luo lomake demokentille',
                            },
                            {
                                'data': templates[1].strip(),
                                'text': 'Lomake (laajempi) ',
                                'expl': 'Laajempi lomake demokentille',
                            },
                        ],
                    },
                ],
            },
        ],
    },
    )


if __name__ == '__main__':
    app.run(
        host='0.0.0.0',
        port=5000,
        debug=False,  # for live reloading, this can be turned on
    )

File: timApp/modules/savearea/js/savearea.ts
/**
 * Defines the client-side implementation of an example plugin (a saveareandrome checker).
 */
import angular, {INgModelOptions} from "angular";
import * as t from "io-ts";
import {GenericPluginMarkup, Info, nullable, PluginBase, withDefault} from "tim/plugin/util";
import {$http} from "tim/util/ngimport";
import {to} from "tim/util/utils";
import {valueDefu} from "tim/util/utils";
import {number} from "../../../static/scripts/jspm_packages/npm/io-ts@1.4.1/lib";

const saveareaApp = angular.module("saveareaApp", ["ngSanitize"]);
export const moduleDefs = [saveareaApp];

const SaveareaMarkup = t.intersection([
    t.partial({
        initword2: t.number,
        inputplaceholder2: nullable(t.number),
        inputstem2: t.string,
    }),
    GenericPluginMarkup,
    t.type({
        // all withDefaults should come here; NOT in t.partial
        autoupdate: withDefault(t.number, 500),
        cols: withDefault(t.number, 20),
    }),
]);
const SaveareaAll = t.intersection([
    t.partial({
        demopisteet: t.number,
    }),
    t.type({
        info: Info,
        markup: SaveareaMarkup,
        preview: t.boolean,
    }),
]);

class SaveareaController extends PluginBase<t.TypeOf<typeof SaveareaMarkup>, t.TypeOf<typeof SaveareaAll>, typeof SaveareaAll> {
    private result?: string;
    private error?: string;
    private isRunning = false;
    private demopisteet = number;
    private modelOpts!: INgModelOptions; // initialized in $onInit, so need to assure TypeScript with "!"

    getDefaultMarkup() {
        return {};
    }

    buttonText() {
        return super.buttonText() || "Tallenna";
    }

    $onInit() {
        super.$onInit();
        this.modelOpts = {debounce: this.autoupdate};
    }

    get autoupdate(): number {
        return this.attrs.autoupdate;
    }

    get inputstem2() {
        return this.attrs.inputstem2 || null;
    }

    get cols() {
        return this.attrs.cols;
    }

    initCode() {
        this.demopisteet = number;
        this.error = undefined;
        this.result = undefined;
    }

    saveText() {
        this.doSaveText(false);
    }

    async doSaveText(nosave: boolean) {
        this.error = "... saving ...";
        this.isRunning = true;
        this.result = undefined;
        const params = {
            input: {
                nosave: false,
                demopisteet: this.demopisteet,
            },
        };

        if (nosave) {
            params.input.nosave = true;
        }
        const url = this.pluginMeta.getAnswerUrl();
        const r = await to($http.put<{web: {result: string, error?: string}}>(url, params));
        this.isRunning = false;
        if (r.ok) {
            const data = r.result.data;
            this.error = data.web.error;
            this.result = data.web.result;
        } else {
            this.error = "Infinite loop or some other error?";
        }
    }

    protected getAttributeType() {
        return SaveareaAll;
    }
}

saveareaApp.component("saveareaRunner", {
    bindings: {
        json: "@",
    },
    controller: SaveareaController,
    template: `
<div class="no-popup-menu">
    <h4 ng-if="::$ctrl.header" ng-bind-html="::$ctrl.header"></h4>
    <p ng-if="::$ctrl.stem">{{::$ctrl.stem}}</p>
    <div class="form-inline">
    <button class="timButton"
            ng-if="::$ctrl.buttonText()"
            ng-click="$ctrl.saveText()">
        {{::$ctrl.buttonText()}}
    </button>
    </div>
    <a href="" ng-if="$ctrl.edited" ng-click="$ctrl.initCode()">{{::$ctrl.resetText}}</a>
    <div ng-if="$ctrl.error" ng-bind-html="$ctrl.error"></div>
    <pre ng-if="$ctrl.result">{{$ctrl.result}}</pre>
    <p ng-if="::$ctrl.footer" ng-bind="::$ctrl.footer" class="plgfooter"></p>
</div>

`,
});